skxray/core/accumulators/correlation/correlation.py

```diff
@@ -0,0 +1,269 @@
+    s = _state
+    # stash the number of pixels in the mask
+    num_pixels = np.bincount(s.label_mask)
+    # Convert from num_levels, num_bufs to lag frames.
+    tot_channels, lag_steps = multi_tau_lags(num_levels, num_bufs)
+
+    # iterate over the images to compute multi-tau correlation
+    for image in image_iterable:
+        # Compute the correlations for all higher levels.
+        level = 0
+
+        # increment buffer
+        s.cur[0] = (1 + s.cur[0]) % num_bufs
+
+        # Put the ROI pixels into the ring buffer.
+        s.buf[0, s.cur[0] - 1] = np.ravel(image)[s.pixel_list]
+        buf_no = s.cur[0] - 1
+        # Compute the correlations between the first level
+        # (undownsampled) frames. This modifies G,
+        # past_intensity, future_intensity,
+        # and img_per_level in place!
+        processing_func(s.buf, s.G, s.past_intensity, s.future_intensity,
+                        s.label_mask, num_bufs, num_pixels, s.img_per_level,
+                        level, buf_no)
+
+        # check whether the number of levels is one, otherwise
+        # continue processing the next level
+        processing = num_levels > 1
+
+        level = 1
+        while processing:
+            if not s.track_level[level]:
+                s.track_level[level] = True
+                processing = False
+            else:
+                prev = (1 + (s.cur[level - 1] - 2) % num_bufs)
+                s.cur[level] = (
+                    1 + s.cur[level] % num_bufs)
+
+                # TODO clean this up. it is hard to understand
+                s.buf[level, s.cur[level] - 1] = ((
+                        s.buf[level - 1, prev - 1] +
+                        s.buf[level - 1, s.cur[level - 1] - 1]
+                    ) / 2
+                )
+
+                # make the track_level zero once that level is processed
+                s.track_level[level] = False
+
+                # call processing_func for each multi-tau level greater
+                # than one. This is modifying things in place. See comment
+                # on previous call above.
+                buf_no = s.cur[level] - 1
+                processing_func(s.buf, s.G, s.past_intensity,
+                                s.future_intensity, s.label_mask, num_bufs,
+                                num_pixels, s.img_per_level, level, buf_no)
+                level += 1
+
+                # Checking whether there is next level for processing
+                processing = level < num_levels
+
+        # If any past intensities are zero, then g2 cannot be normalized at
+        # those levels. This if/else code block is basically preventing
+        # divide-by-zero errors.
+        if len(np.where(s.past_intensity == 0)[0]) != 0:
+            g_max = np.where(s.past_intensity == 0)[0][0]
+        else:
+            g_max = s.past_intensity.shape[0]
+
+        # Normalize g2 by the product of past_intensity and future_intensity
+        g2 = (s.G[:g_max] /
```

Comment:
There is little gain in this.  As per this question, I benchmarked the time savings for computing versus not computing g2 here.  The notebook can be found [here] (https://gist.github.com/ericdill/f6b54927b184b6a47be9) and the results show that there is about a ~2% speed up in not computing g2 for very small ROIs and no noticeable difference if the ROIs are very large.